File: depth_fisrt_search/depth_fisrt_search.java
import java.util.ArrayList;
import java.util.List;

public class Node {
    String name;
    List<Node> children = new ArrayList<>();

    public Node(String name) {
        this.name = name;
    }

    public Node addChild(String name) {
        Node child = new Node(name);
        children.add(child);
        return this;
    }

    public List<String> depthFirstSearch(List<String> array) {
        array.add(this.name);
        for (Node child : children) {
            child.depthFirstSearch(array);
        }
        return array;
    }

    public static void main(String[] args) {
        Node root = new Node("A");
        root.addChild("B").addChild("C");
        root.children.get(0).addChild("D");

        List<String> result = root.depthFirstSearch(new ArrayList<>());
        System.out.println(result);
    }
}

File: depth_fisrt_search/depth_fisrt_search.go
package main

import (
	"fmt"
)

type Node struct {
	Name     string
	Children []*Node
}

func (n *Node) AddChild(name string) *Node {
	child := &Node{Name: name}
	n.Children = append(n.Children, child)
	return n
}

func (n *Node) DepthFirstSearch(array *[]string) []string {
	*array = append(*array, n.Name)
	for _, child := range n.Children {
		child.DepthFirstSearch(array)
	}
	return *array
}

func main() {
	root := &Node{Name: "A"}
	root.AddChild("B").AddChild("C")
	root.Children[0].AddChild("D")

	result := []string{}
	root.DepthFirstSearch(&result)
	fmt.Println(result)
}
